rasa/utils/tensorflow/data_generator.py

```diff
@@ -0,0 +1,475 @@
+        # for nonexistent inputs. To save calculation we filter this features before
+        # input to tf methods.
+        number_of_features = array_of_array_of_sparse[0][0].shape[-1]
+        array_of_array_of_sparse = RasaDataGenerator._filter_out_fake_inputs(
+            array_of_array_of_sparse
+        )
+        if not array_of_array_of_sparse:
+            # create empty array with appropriate last dims
+            return [
+                np.empty((0, 3), dtype=np.int64),
+                np.array([], dtype=np.float32),
+                np.array([0, 0, number_of_features], dtype=np.int64),
+            ]
+
+        # we need to make sure that the matrices are coo_matrices otherwise the
+        # transformation does not work (e.g. you cannot access x.row, x.col)
+        if not isinstance(array_of_array_of_sparse[0][0], scipy.sparse.coo_matrix):
+            array_of_array_of_sparse = [
+                [x.tocoo() for x in array_of_sparse]
+                for array_of_sparse in array_of_array_of_sparse
+            ]
+
+        dialogue_len = [
+            len(array_of_sparse) for array_of_sparse in array_of_array_of_sparse
+        ]
+        combined_dialogue_len = sum(dialogue_len)
+        max_seq_len = max(
+            [
+                x.shape[0]
+                for array_of_sparse in array_of_array_of_sparse
+                for x in array_of_sparse
+            ]
+        )
+        # get the indices of values
+        indices = np.hstack(
+            [
+                np.vstack(
+                    [sum(dialogue_len[:i]) + j * np.ones_like(x.row), x.row, x.col]
+                )
+                for i, array_of_sparse in enumerate(array_of_array_of_sparse)
+                for j, x in enumerate(array_of_sparse)
+            ]
+        ).T
+
+        data = np.hstack(
+            [
+                x.data
+                for array_of_sparse in array_of_array_of_sparse
+                for x in array_of_sparse
+            ]
+        )
+
+        shape = np.array((combined_dialogue_len, max_seq_len, number_of_features))
+
+        return [
+            indices.astype(np.int64),
+            data.astype(np.float32),
+            shape.astype(np.int64),
+        ]
+
+    @staticmethod
+    def _filter_out_fake_inputs(
+        array_of_array_of_features: FeatureArray,
+    ) -> Union[List[List[np.ndarray]], List[List[scipy.sparse.spmatrix]]]:
+        return list(
+            filter(
+                # filter empty lists created by another filter
+                lambda x: len(x) > 0,
+                [
+                    # filter all the "fake" inputs, we know the input is "fake",
+                    # when sequence dimension is `0`
+                    list(filter(lambda x: x.shape[0] > 0, array_of_features))
+                    for array_of_features in array_of_array_of_features
+                ],
+            )
+        )
+
+
+class IncreasingBatchSizeDataGenerator(RasaDataGenerator):
```

Comment:
Q: Do we need an extra class for `IncreasingBatchSizeDataGenerator`? Can't we do something like - if passed `batch_size` is a `list` rather than `int`, we call `linearly_increasing_batch_size` to get the next batch size?

Reply:
Yeah, I guess we could do that, like the idea actually 👍

Reply:
Should we keep the name `IncreasingBatchSizeDataGenerator` or would you go with something else?

Reply:
I'd still use increasing, sometimes it can be increasing by `0` 😂

Reply:
I would suggest just having one class named `BatchDataGenerator`. Increasing batch is a property of that class depending on the value of `batch_size`. What do you think?

Reply:
let's then add `Rasa` so that we know that it uses custom rasa data structure and that it is not general class. Similar separation we have with our models and layers, where layers without `Rasa...` are general reusable layers by any model

Reply:
Let's keep the `RasaDataGenerator` with the more generic methods as I will create a second data generator that is able to load data from files. 
I merged the `FixedBatchSizeDataGenerator` and the `IncreasingBatchSizeDataGenerator` into one class. It is currently still called `IncreasingBatchSizeDataGenerator`. 
`BatchDataGenerator` seems to be too generic I think. In the end all data generators create some kind of batches.

Reply:
updated the name to `RasaBatchDataGenerator`